aten/src/ATen/native/cpu/group_norm_kernel.cpp

```diff
@@ -393,19 +392,30 @@ void GroupNormKernelImpl(
     Tensor& Y,
     Tensor& mean,
     Tensor& rstd) {
+  const bool mixed_type = is_mixed_type(X, gamma, beta);
   switch (X.suggest_memory_format()) {
     case at::MemoryFormat::Contiguous: {
       AT_DISPATCH_FLOATING_TYPES_AND(ScalarType::BFloat16, X.scalar_type(), "GroupNormKernelImpl", [&]() {
-        GroupNormKernelImplInternal<scalar_t>(
-            X, gamma, beta, N, C, HxW, group, static_cast<scalar_t>(eps), Y, mean, rstd);
+        if (mixed_type) {
+          GroupNormKernelImplInternal<BFloat16, float>(
+              X, gamma, beta, N, C, HxW, group, static_cast<scalar_t>(eps), Y, mean, rstd);
+        } else {
```

Comment:
Hmm, this will add this if statement for all datatypes, but it makes sense only for float16, isn't it? Can we use a constexpr or something to restrict this one only for BFloat16 scalar type?

Reply:
Yes, users can create mixed dtype cases for other data types than bfloat16+float, even though these are not typical. Maybe a simpler way to fix it is to move the dtype check of bfloat16 and mixed type check out and before `AT_DISPATCH_FLOATING_TYPES_AND`.

Reply:
changed the mixed_dtype pattern into the following:
```C++
        using param_t = at::opmath_type<scalar_t>;
        if (mixed_type) {
          GroupNormKernelImplInternal<scalar_t, param_t>(
              X, gamma, beta, N, C, HxW, group, static_cast<scalar_t>(eps), Y, mean, rstd);
        } else {
          GroupNormKernelImplInternal<scalar_t, scalar_t>(
              X, gamma, beta, N, C, HxW, group, static_cast<scalar_t>(eps), Y, mean, rstd);
        }
```
We will enable `mixed_dtype` with float16 in near future, so `mixed_dtype` will be like:
* bfloat16 + float32
* float16 + float32

And for float16, we won't have to do anything special, just replacing the `AT_DISPATCH_FLOATING_TYPES_AND` will do the entire job.